pallets/pallet-asset-swap/src/xcm/trade.rs

```diff
@@ -109,56 +155,147 @@ mod swap_pair_remote_asset {
 				Error::FailedToTransactAsset("Failed to convert swap pair asset ID into required version.")
 			}).ok()?;
-			let weight = weight.min(self.0);
+
+			let weight = weight.min(self.remaining_weight);
 			let amount = WeightToFee::weight_to_fee(&weight);
-			self.0 -= weight;
-			self.1 = self.1.saturating_sub(amount);
-			let amount: u128 = amount.saturated_into();
+
+			self.consumed_xcm_hash = None;
+			self.remaining_fungible_balance = self
+				.remaining_fungible_balance
+				.saturating_sub(self.remaining_fungible_balance);
+			self.remaining_weight = self.remaining_weight.saturating_sub(weight);
+
 			if amount > 0 {
-				log::trace!(target: LOG_TARGET, "refund amount {:?}", (swap_pair_remote_asset_v3, amount));
+				log::trace!(target: LOG_TARGET, "Refund amount {:?}", (swap_pair_remote_asset_v3, amount));
 				Some((swap_pair_remote_asset_v3, amount).into())
 			} else {
 				log::trace!(target: LOG_TARGET, "No refund");
 				None
 			}
 		}
 	}
+
+	// Move any unused asset from the swap pool account to the specified account,
+	// and update the remote balance with the difference since we know we control
+	// the full amount on the remote location.
+	impl<T, WeightToFee, FeeDestinationAccount> Drop
+		for UsingComponentsForSwapPairRemoteAsset<T, WeightToFee, FeeDestinationAccount>
+	where
+		T: Config,
+		FeeDestinationAccount: Get<T::AccountId>,
+	{
+		fn drop(&mut self) {
+			log::trace!(target: LOG_TARGET, "Drop with remaining {:?}", (self.consumed_xcm_hash, self.remaining_fungible_balance, self.remaining_weight, &self.swap_pair));
+			match (self.remaining_fungible_balance, &self.swap_pair) {
+				(remaining_balance, Some(swap_pair)) if remaining_balance > Zero::zero() => {
+					let Ok(remaining_balance_as_local_currency) = LocalCurrencyBalanceOf::<T>::try_from(remaining_balance).map_err(|e| {
+						log::error!(target: LOG_TARGET, "Failed to convert remaining balance {:?} to local currency balance", remaining_balance);
+						e
+					}) else { return; };
+
+					// No error should ever be thrown from inside this block.
+					let _ = <T::LocalCurrency as Mutate<T::AccountId>>::transfer(
+						&swap_pair.pool_account,
+						&FeeDestinationAccount::get(),
+						remaining_balance_as_local_currency,
+						Preservation::Expendable,
+					).map_err(|e| {
+						log::error!(target: LOG_TARGET, "Failed to transfer unused balance {:?} from swap pair pool account {:?} to specified account {:?}", remaining_balance_as_local_currency, swap_pair.pool_account, FeeDestinationAccount::get());
+						e
+					});
+
+					// No error should ever be thrown from inside this block.
+					SwapPair::<T>::mutate(|entry| {
+						let Some(entry) = entry.as_mut() else {
+							log::error!(target: LOG_TARGET, "Stored swap pair should not be None but it is.");
+							return;
+						};
+						entry.remote_asset_balance = entry
```

Comment:
I think a * is missing, no?

Reply:
It's a `u128`. I also was assuming so, but apparently you cannot dereference a `u128`. And since the tests are passing, I assume we are changing the storage directly here 😄